torch/csrc/jit/autodiff.cpp

```diff
@@ -473,10 +479,63 @@ static std::vector<Value*> gradientForNode(Node* node, ArrayRef<Value*> grad_val
     }
     throw std::runtime_error(std::string("failed to differentiate `") + node->kind().toDisplayString() + "`");
   };
+  // Write gradient in Python and compile it to JIT graphs.
+  // For example, aten::mul() should be defined as follows
+  // def forward(x, y):
+  //     return x*y, (x, y)
+  // def backward(ctx, grad_output):
+  //     x, y = ctx
+  //     return y * grad_output, x * grad_output
+  // Note that ctx is a tuple that carries all input/intermediate results needed
+  // in backward from forward pass.
+  // grad_values(a.k.a gradOutputs) propagated through node->owningGraph() in reversed order.
+  // The forward graph should be inserted **after** the node being replaced,
+  // so that we don't traverse it infinite times.
+  // The output of compiled forward graph is [real_outputs, ctx]
+  // The input of compiled backward graph is [ctx, grad_values]
+  // We run LowerSimpleTuples afterwards to elmininate all tuples generated in this process.
+  // The original node will be cleaned up later in EliminateDeadCode
+  const auto build_script_grad = [node](const ArrayRef<Value*>& grads, const std::string& source) -> std::vector<Value*> {
+    // Compile the python code to a script module
+    auto cu = std::make_shared<script::Module>();
+    script::defineMethodsInModule(cu, source, script::nativeResolver, nullptr);
+    auto graph = node->owningGraph();
+
+    // Use forward graph to replace node in grad_desc.f
+    value_list new_outputs;
+    {
+      WithInsertPoint guard(node->next());
+      auto fw_graph = cu->find_method("forward")->graph();
+      new_outputs = script::inlineCallTo(*graph, *fw_graph, node->inputs());
+      for (size_t i = 0; i < node->outputs().size(); ++i) {
+        new_outputs.at(i)->setType(node->outputs()[i]->type());
+        new_outputs.at(i)->replaceAllUsesWith(node->outputs()[i]);
+      }
+    }
+
+    // Use backward graph to construct reverse_block
+    auto bwgraph = cu->find_method("backward")->graph();
+    auto grad_vec = grads.vec();
+    auto it = grad_vec.begin();
+    grad_vec.insert(it, new_outputs.back());
+    ArrayRef<Value*> grad(grad_vec);
+    auto grad_inputs = script::inlineCallTo(*graph, *bwgraph, grad);
+    return grad_inputs;
+  };
+
   if (!isDifferentiable(node)) {
     throw std::runtime_error(std::string("differentiation of ") + node->kind().toDisplayString() + " "
                              "is not supported, or it is missing necessary type information");
   }
+  // If defined using Torchscript, use it instead of symbolic
+  // symbolic_script is a map<schema, python_code_str> defined in symbolic_script.h
+  for (auto& it : symbolic_scripts) {
+    if (node->matches(it.first.c_str())) {
```

Comment:
the const char* passed to matches must be a literal string, it cannot be dynamically allocated.
we will have to figure out another way to do this